golemfactory/yagna-service-erigon
# Authentication to Erigon RPC via nginx

[APPS-127](https://golemproject.atlassian.net/browse/APPS-127)

**Prerequisities**
- `htpasswd tool` is required (👉 Install with `sudo apt-get install apache2-utils`)

**How**
Instead of creating .htpasswd file by it's own (with code) it spawns `htpasswd` tool to write file. Flag `-c` allows to reset the password if the user already exists (which is our case).

**Test it**
- `rm ~/.local/share/ya-erigon-runtime/ya-erigon-runtime.json`
- `cargo build` (in `ya-erigon-runtime` directory)
- start the runtime to generate a config file (in `ya-erigon-runtime` directory)
`ya-runtime-dbg --runtime ./target/debug/ya-erigon-runtime --workdir ./target/debug`

- modify the config file to fill your needs (`cat ~/.local/share/ya-erigon-runtime/ya-erigon-runtime.json`)
```json
{
  "public_addr": "http://erigon.localhost:8545",
  "data_dir": "/data/turbo-geth",
  "passwd_tool_path": "/usr/bin/htpasswd",
  "passwd_file_path": "<YOUR HTPASSWD FILE PATH HERE>", 👈 ❗ 
  "password_default_length": 15
}
```

- make sure `tg` & `rpcdaemon` binaries are present in `ya-erigon-runtime/target/debug` directory
- start the runtime again `ya-runtime-dbg --runtime ./target/debug/ya-erigon-runtime --workdir ./target/debug`

(this starts the debugger - write `something` `ENTER` which exposes the erigon status:
⚠️ Indentation added for 👁️ visibility
```
[INFO] Entering prompt, press C-d to exit
bash ▶ ddd
[0] output for command: RunProcess { bin: "/bin/bash", args: ["bash", "-c", "ddd"], work_dir: "", stdout: None, stderr: None }. \
 ERIGON: {
    "auth": {
         "password":"RLiEm1K0rJEj3F0",
         "user":"erigolem"
    },
    "status":"running",
    "url":"http://erigon.localhost:8545"
}
```


**Verify**
(In the new terminal window)
- check the erigon has started and it respond to rpc (localhost:8545)
- check the presented username & password are correct
`htpasswd -vb <PATH TO HTPASSWD FILE> erigolem <PASSWORD FROM THE DEBUGGER>`

Go back to the debugger and stop the service (`ctrl-d`)
